### DISCUSSION TITLE
Formalize folder naming conventions

### DISCUSSION THREAD

**Discussion by mattbrailsford** [umbraco/Umbraco-CMS#10638] [Features and ideas]
Given .NET Core is now case sensitive it would be good if we could formalize the naming convention of folders in the project as currently, it appears to be a mix

![image](https://user-images.githubusercontent.com/527305/125085246-44f33980-e0c2-11eb-814c-e2e44035606c.png)

From a discussion on another issue it was suggested everything in the content root should probably be first upper where as anything that is in the wwwroot should be all lower.

It would be good if there was some actual standard to this.

**Comment by mattbrailsford** [comment]
Another good example, why is Umbraco's own language file config folder `config/lang` yet all plugins lang folders are `Lang`?

![image](https://user-images.githubusercontent.com/527305/142194331-8b5c025d-7aa3-4a6c-8d11-f250d4438cdb.png)

**Comment by leekelleher** [comment]
Back in my earlier web days, (circa. '99), I recall reading a UK Gov guideline about their file and folder name conventions, I've never been able to find the document since, (probably lost in the ether), but the general gist was **kebab-case (lowercase)** everything, particularly if those files were publicly (web) accessible.

> Update, found this: <https://insidegovuk.blog.gov.uk/url-standards-for-gov-uk/>
> They don't say kebab-case verbatim, but suggest it in point 3 about using dashes.

**Comment by p-m-j** [comment]
We should also formalize, document, and split folders that contain environment specific files e.g. logs from those which should be version controlled & published!

**Comment by markadrake** [comment]
I bounce between Windows and macOS frequently and this gives me some heartache in command line and terminal windows respectively. The ideal solution (in my opinion) is:

- all lowercase
- and-use-kebabs-too

_I won't go into how I reconcile this file/folder naming pattern with HTML, CSS, and JavaScript... I just have to deal with it._

**Comment by ronaldbarendse** [comment]
We did have an internal discussion about naming/location conventions just before releasing v9, as I noticed the same while updating the ImageSharp.Web cache directory: https://github.com/umbraco/Umbraco-CMS/pull/10884 (see my related note at the bottom of the PR description). The proposed changes were too breaking or too late to implement before the final release, as it would break a lot of existing local developer setups (ignored files getting visible again, require manually moving files/directories), break existing packages and require changes to the hosting configuration.

I agree that everything in the normal project structure should use Pascal Case (as is the Microsoft/C#/.NET/MVC default) and only front-end files in the `wwwroot` folder should be lowercased (so we can use lowercase URLs for them). One exception might currently be the `App_Plugins` folder though 🤔 

This is my proposed structure:

```
\App_Data - writable environment specific data (git ignored)
  \Logs - Serilog JSON file sink
  \TEMP - NuCache, Examine indexes, temporary file uploads, etc.
  \*.sdf/mdf - development database files
\App_Plugins - publicly served
\Umbraco - Umbraco specific files
  \Lang - default language files
  \Data - writable Umbraco data
    \Packages - repository for packages created from the back-office (should be moved to the database)
    \Deploy - Deploy data/marker files
      \Revision - Deploy artifacts
    \Models - generated models from Models Builder
  \Views
    \PartialViewMacros\Templates - default templates used for macros
    \BackOffice - default views for back-office (upgrade/preview)
    \Install - default views for install
    \Website - default views for website (no nodes/not found)
\Views - project views (only ship a default _ViewImports.cshtml, remove Grid/Block List views)
\wwwroot
  \media - writable media store (might make sense to move this to App_Data\Media and publicly serve using UseStaticFiles)
  \umbraco - Umbraco assets required by the backoffice
```

**Reply by p-m-j** [reply]
Still prefer /Environment to /App_Data :p

Any reason not to just have App_plugins be wwwroot/plugins v10+?

**Reply by leekelleher** [reply]
App_Plugins is feeling like a hangover from ASP.NET Framework. 🍻🥴

**Reply by benjaminc** [reply]
I would love to see this as part of v10.  My proposed directories would be similar to @ronaldbarendse but all kebab-case across the board.  It would just make it easier when developing on Linux, as well as deploying to Linux.  Where possible, allow for the older folder names to still be used so that not everyone has to change again after just upgrading to v9, but make sure the community is well aware these are going away and aren't the standard anymore, and then remove them in the next LTS major after v10 (v14 I think?).

```
\local - writable area for files specific to this instance/environment (git ignored)
    \logs - Serilog JSON file sink
    \temp - nu-cache, examine, temporary file uploads, etc.
    \*.sdf/mdf - development database files
\plugins - publicly served
\umbraco - Umbraco specific files
    \lang - default language files
    \data - writable Umbraco data
        \packages - repository for packages created from the back-office (should be moved to the database)
        \deploy - Deploy data/marker files
            \revision - Deploy artifacts
        \models - generated models from Models Builder
    \views
        \partial-view-macros\templates - default templates used for macros
        \back-office - default views for back-office (upgrade/preview)
        \install - default views for install
        \website - default views for website (no nodes/not found)
\views - project views (only ship a default _ViewImports.cshtml, remove Grid/Block List views)
\wwwroot
    \media - writable media store (might make sense to move this to App_Data\Media and publicly serve using UseStaticFiles)
    \umbraco - Umbraco assets required by the backoffice
```
